akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala

```diff
@@ -241,24 +252,32 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
             Behaviors.same
           }
 
-        case NodeRemoved(address) ⇒
+        case NodeRemoved(uniqueAddress) ⇒
           // ok to update from several nodes but more efficient to try to do it from one node
           if (cluster.state.leader.contains(cluster.selfAddress)) {
-            nodesRemoved(Set(address))
+            ctx.log.debug(s"Leader node observed removed address [{}]", uniqueAddress)
+            nodesRemoved(Set(uniqueAddress))
           } else Behaviors.same
 
         case RemoveTick ⇒
           // ok to update from several nodes but more efficient to try to do it from one node
           if (cluster.state.leader.contains(cluster.selfAddress)) {
-            val allAddressesInState: Set[Address] = state.map.entries.flatMap {
-              case (_, values) ⇒
+            val allAddressesInState: Set[UniqueAddress] = state.map.entries.flatMap {
+              case (_, entries) ⇒
                 // don't care about local (empty host:port addresses)
-                values.collect { case ref if ref.path.address.hasGlobalScope ⇒ ref.path.address }
+                entries.collect {
+                  case entry if entry.ref.path.address.hasGlobalScope ⇒
+                    entry.uniqueAddress(setup.selfUniqueAddress)
+                }
             }(collection.breakOut)
-            val clusterAddresses = cluster.state.members.map(_.address)
-            val diff = allAddressesInState diff clusterAddresses
-            if (diff.isEmpty) Behavior.same
-            else nodesRemoved(diff)
+            val clusterAddresses = cluster.state.members.map(_.uniqueAddress)
+            val notInCluster = allAddressesInState -- clusterAddresses
+
+            if (notInCluster.isEmpty) Behavior.same
+            else {
+              ctx.log.debug("Leader node cleanup tick, removed nodes: [{}]", notInCluster.mkString(","))
+              nodesRemoved(notInCluster)
+            }
```

Comment:
In the ticket I wrote:

> If a node with same hostname:port joins again there is risk that we miss the removal of some refs that belongs to the old incarnation. I think we need a data structure that also includes the owning system's uid for each ref.
>
> Another scenario to consider is when a registered actor is terminated, but the node is still up and running. That removal can also be "lost" and would result in a dead actor ref lingering around.

This PR only takes care of the first scenario. The second scenario is still remaining. Let's create a new issue for that.

Reply:
Created #24887